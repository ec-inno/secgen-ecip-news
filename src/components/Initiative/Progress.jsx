import React from 'react';
import { useTranslation } from 'react-i18next';

/**
 * When a given status is marked as "assumed", this means that no actual data
 * has been found during the development of the component and named has been
 * thought of. Also, the fact that these are assumed means there are no date
 * fields matched to them currently.
 * Meaning, on "SUBMITTED" and "ANSWERED" we won't display dates.
 */
const Progress = ({ initiative }) => {
  const { t } = useTranslation();

  const supportedStatuses = [
    'REGISTERED',
    'COLLECTION_CLOSED',
    'VERIFICATION', // assumed
    'SUBMITTED', // assumed
    'ANSWERED',
    'CONDITIONS_NOT_FULFILLED',
    'WITHDRAWN_BY_ORGANISER',
  ];

  // In few cases input cannot be handled.
  if (!initiative || !initiative.status) return '';

  if (
    initiative.status &&
    supportedStatuses.indexOf(initiative.status) === -1
  ) {
    return (
      <>
        <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
        <p className="ecl-u-type-paragraph">
          {t('Given status cannot be mapped to currently supported timeline.')}
        </p>
      </>
    );
  }

  // Prepare information about some fields.
  const registrationDate = initiative.registrationDate
    ? getDateFormatted(initiative.registrationDate)
    : '';

  const collectionDeadline =
    initiative.searchEntry && initiative.searchEntry.deadlineForCollection
      ? getDateFormatted(initiative.searchEntry.deadlineForCollection)
      : '';

  switch (initiative.status) {
    case 'REGISTERED': {
      return (
        <>
          <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
          <ol className="ecl-timeline" data-ecl-timeline="true">
            {registrationDate ? (
              <li className="ecl-timeline__item ecl-timeline__item--current">
                <div className="ecl-timeline__label">{t('Registered')}</div>
                <div className="ecl-timeline__content">
                  {t('Deadline')}
                  {': '}
                  {registrationDate}
                </div>
              </li>
            ) : (
              <li className="ecl-timeline__item ecl-timeline__item--current">
                <div className="ecl-timeline__label">{t('Registered')}</div>
                <div className="ecl-timeline__content">
                  {t('Current status')}
                </div>
              </li>
            )}
            {collectionDeadline ? (
              <li className="ecl-timeline__item">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
                <div className="ecl-timeline__content">
                  {t('Deadline')}
                  {': '}
                  {collectionDeadline}
                </div>
              </li>
            ) : (
              <li className="ecl-timeline__item ecl-timeline__item--disabled">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
              </li>
            )}
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">
                {t('Collection closed')}
              </div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Verification')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Submitted')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Answered')}</div>
            </li>
          </ol>
        </>
      );
    }

    case 'COLLECTION_CLOSED': {
      return (
        <>
          <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
          <ol className="ecl-timeline" data-ecl-timeline="true">
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">{t('Registered')}</div>
              {registrationDate && (
                <div className="ecl-timeline__content">{registrationDate}</div>
              )}
            </li>
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">
                {t('Collection ongoing')}
              </div>
            </li>
            {collectionDeadline ? (
              <li className="ecl-timeline__item ecl-timeline__item--current">
                <div className="ecl-timeline__label">
                  {t('Collection closed')}
                </div>
                <div className="ecl-timeline__content">
                  {collectionDeadline}
                </div>
              </li>
            ) : (
              <li className="ecl-timeline__item ecl-timeline__item--current">
                <div className="ecl-timeline__label">
                  {t('Collection closed')}
                </div>
                <div className="ecl-timeline__content">
                  {t('Current status')}
                </div>
              </li>
            )}
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Verification')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Submitted')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Answered')}</div>
            </li>
          </ol>
        </>
      );
    }

    case 'VERIFICATION': {
      return (
        <>
          <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
          <ol className="ecl-timeline" data-ecl-timeline="true">
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">{t('Registered')}</div>
              {registrationDate && (
                <div className="ecl-timeline__content">{registrationDate}</div>
              )}
            </li>
            {collectionDeadline ? (
              <li className="ecl-timeline__item">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
                <div className="ecl-timeline__content">
                  {collectionDeadline}
                </div>
              </li>
            ) : (
              <li className="ecl-timeline__item">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
              </li>
            )}
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">
                {t('Collection closed')}
              </div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--current">
              <div className="ecl-timeline__label">{t('Verification')}</div>
              <div className="ecl-timeline__content">{t('Current status')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Submitted')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Answered')}</div>
            </li>
          </ol>
        </>
      );
    }

    case 'SUBMITTED': {
      return (
        <>
          <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
          <ol className="ecl-timeline" data-ecl-timeline="true">
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">{t('Registered')}</div>
              {registrationDate && (
                <div className="ecl-timeline__content">{registrationDate}</div>
              )}
            </li>
            {collectionDeadline ? (
              <li className="ecl-timeline__item">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
                <div className="ecl-timeline__content">
                  {collectionDeadline}
                </div>
              </li>
            ) : (
              <li className="ecl-timeline__item">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
              </li>
            )}
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">
                {t('Collection closed')}
              </div>
            </li>
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">{t('Verification')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--current">
              <div className="ecl-timeline__label">{t('Submitted')}</div>
              <div className="ecl-timeline__content">{t('Current status')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--disabled">
              <div className="ecl-timeline__label">{t('Answered')}</div>
            </li>
          </ol>
        </>
      );
    }

    case 'ANSWERED': {
      return (
        <>
          <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
          <ol className="ecl-timeline" data-ecl-timeline="true">
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">{t('Registered')}</div>
              {registrationDate && (
                <div className="ecl-timeline__content">{registrationDate}</div>
              )}
            </li>
            {collectionDeadline ? (
              <li className="ecl-timeline__item">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
                <div className="ecl-timeline__content">
                  {collectionDeadline}
                </div>
              </li>
            ) : (
              <li className="ecl-timeline__item">
                <div className="ecl-timeline__label">
                  {t('Collection ongoing')}
                </div>
              </li>
            )}
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">
                {t('Collection closed')}
              </div>
            </li>
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">{t('Verification')}</div>
            </li>
            <li className="ecl-timeline__item">
              <div className="ecl-timeline__label">{t('Submitted')}</div>
            </li>
            <li className="ecl-timeline__item ecl-timeline__item--current">
              <div className="ecl-timeline__label">{t('Answered')}</div>
              <div className="ecl-timeline__content">{t('Current status')}</div>
            </li>
          </ol>
        </>
      );
    }

    case 'CONDITIONS_NOT_FULFILLED':
    case 'WITHDRAWN_BY_ORGANISER': {
      return (
        <>
          <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
          <p className="ecl-u-type-paragraph">
            {t("Initiative hasn't reached registration phase.")}
          </p>
        </>
      );
    }

    default: {
      return (
        <>
          <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
          <p className="ecl-u-type-paragraph">{t('N/A')}</p>
        </>
      );
    }
  }
};

export default Progress;
