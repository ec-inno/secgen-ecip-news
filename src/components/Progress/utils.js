export const steps = [
  'REGISTERED',
  'COLLECTION_START_DATE',
  'ONGOING',
  'COLLECTION_EARLY_CLOSURE',
  'CLOSED',
  'VERIFICATION',
  'SUBMITTED',
  'ANSWERED',
];

/**
 * Reorder progress stages to match order of `steps`.
 * @param {Array<Object>} progress
 * @param {String} progress.name
 * @param {Boolean} progress.active
 * @param {Date} progress.date
 */
export const getStages = progress => {
  const stages = [];

  steps.forEach(step => {
    const match = progress.find(item => item.name === step);
    if (match) stages.push(match);
  });

  return stages;
};
