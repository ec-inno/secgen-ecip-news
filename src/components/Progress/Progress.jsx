import React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import upperCaseFirstChar from '@eci/utils/upperCaseFirstChar';

import { steps } from './utils';

const Progress = ({ progress, dateStart, dateEnd }) => {
  const { t } = useTranslation();

  if (!progress || progress.length === 0) {
    return (
      <>
        <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
        <p className="ecl-u-type-paragraph">
          {t('Given status cannot be mapped to currently supported timeline.')}
        </p>
      </>
    );
  }

  const stages = [];
  const timeline = [];

  // Reorder progress stages to match order of `steps`.
  steps.forEach(step => {
    const match = progress.find(item => item.name === step);
    if (match) stages.push(match);
  });

  // Build the timeline elements.
  stages.forEach((stage, key) =>
    timeline.push(
      <li
        key={key}
        className={
          stage.active
            ? 'ecl-timeline__item ecl-timeline__item--current'
            : 'ecl-timeline__item'
        }
      >
        <div className="ecl-timeline__label">
          {upperCaseFirstChar(stage.name)}
        </div>
        {stage.date && (
          <div className="ecl-timeline__content">{stage.date}</div>
        )}
      </li>
    )
  );

  return (
    <>
      <h3 className="ecl-u-type-heading-3">{t('Initiative progress')}</h3>
      <ol className="ecl-timeline" data-ecl-timeline="true">
        {timeline}
      </ol>
      {dateStart && (
        <p className="ecl-u-type-paragraph-s ecl-u-type-bold">
          {t('Collection start date')}
          {': '}
          {dateStart}
        </p>
      )}
      {dateEnd && (
        <p className="ecl-u-type-paragraph-s ecl-u-type-bold">
          {t('Collection closed earlier by the organisers')}
          {': '}
          {dateEnd}
        </p>
      )}
    </>
  );
};

Progress.propTypes = {
  progress: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      active: PropTypes.bool,
      date: PropTypes.string,
    })
  ),
  dateStart: PropTypes.string,
  dateEnd: PropTypes.string,
};

export default Progress;
