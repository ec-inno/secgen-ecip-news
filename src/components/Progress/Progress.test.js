import React from 'react';
import renderer from 'react-test-renderer';

import Progress from './Progress';

describe('Progress utilities', () => {
  it('steps supported are public', () => {
    const { steps } = require('./utils');
    expect(steps).not.toBe(undefined);
    expect(Array.isArray(steps)).toBe(true);
    expect(steps.length > 0).toBe(true);
    expect(steps).toMatchSnapshot();
  });

  it('contains information about collection start and early closure dates in specific places', () => {
    const { steps } = require('./utils');
    expect(steps[1]).toBe('COLLECTION_START_DATE');
    expect(steps[3]).toBe('COLLECTION_EARLY_CLOSURE');
  });

  it('getStages() reorders items if not in correct order', () => {
    const progress = [
      {
        name: 'ONGOING',
        active: true,
        date: '09/09/2019',
      },
      {
        name: 'REGISTERED',
        active: false,
        date: '11/08/2019',
      },
    ];
    const { getStages } = require('./utils');

    const stages = getStages(progress);
    expect(stages[0].name).toBe('REGISTERED');
    expect(stages[1].name).toBe('ONGOING');
  });
});

describe('Progress', () => {
  it('no input: provides feedback for unsupported timeline', () => {
    const tree = renderer.create(<Progress />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders timeline in right order regardless of input order', () => {
    const props = {
      progress: [
        {
          name: 'ONGOING',
          active: true,
          date: '09/09/2019',
        },
        {
          name: 'REGISTERED',
          active: false,
          date: '11/08/2019',
        },
      ],
    };

    const tree = renderer.create(<Progress {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('collection start date can be passed as a parameter', () => {
    const props = {
      progress: [
        {
          name: 'ONGOING',
          active: true,
          date: '09/09/2019',
        },
        {
          name: 'REGISTERED', // Yes, spec says this step will be given conditionally, but we still add it to ensure registration start date comes right after registration.
          active: false,
          date: '11/08/2019',
        },
      ],
      dateCollectionStart: '11/08/2019',
    };

    const tree = renderer.create(<Progress {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('collection start date can be passed as a step', () => {
    const props = {
      progress: [
        {
          name: 'ONGOING',
          active: true,
          date: '09/09/2019',
        },
        {
          name: 'REGISTERED',
          active: false,
          date: '11/08/2019',
        },
        {
          name: 'COLLECTION_START_DATE',
          active: false,
          date: '22/09/2019',
        },
      ],
    };

    const tree = renderer.create(<Progress {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('collection early closure date can be passed as a parameter', () => {
    const props = {
      progress: [
        {
          name: 'ONGOING',
          active: true,
          date: '09/09/2019',
        },
        {
          name: 'REGISTERED',
          active: false,
          date: '11/08/2019',
        },
      ],
      dateCollectionEarlyClosure: '11/08/2019',
    };

    const tree = renderer.create(<Progress {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('collection early closure date can be passed as a step', () => {
    // Sample props are not to be passed like this from the service, but we double-check rendered order.
    const props = {
      progress: [
        {
          name: 'ONGOING',
          active: true,
          date: '09/09/2019',
        },
        {
          name: 'CLOSED',
          active: false,
          date: '11/08/2019',
        },
        {
          name: 'COLLECTION_EARLY_CLOSURE',
          active: false,
          date: '11/08/2019',
        },
      ],
    };

    const tree = renderer.create(<Progress {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
