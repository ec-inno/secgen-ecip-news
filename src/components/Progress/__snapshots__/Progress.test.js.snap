// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress collection early closure date can be passed as a parameter 1`] = `
Array [
  <h3
    className="ecl-u-type-heading-3"
  >
    Initiative progress
  </h3>,
  <ol
    className="ecl-timeline"
    data-ecl-timeline="true"
  >
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Registered
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item ecl-timeline__item--current"
    >
      <div
        className="ecl-timeline__label"
      >
        Ongoing
      </div>
      <div
        className="ecl-timeline__content"
      >
        09/09/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Collection early closure
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
  </ol>,
]
`;

exports[`Progress collection early closure date can be passed as a step 1`] = `
Array [
  <h3
    className="ecl-u-type-heading-3"
  >
    Initiative progress
  </h3>,
  <ol
    className="ecl-timeline"
    data-ecl-timeline="true"
  >
    <li
      className="ecl-timeline__item ecl-timeline__item--current"
    >
      <div
        className="ecl-timeline__label"
      >
        Ongoing
      </div>
      <div
        className="ecl-timeline__content"
      >
        09/09/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Collection early closure
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Closed
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
  </ol>,
]
`;

exports[`Progress collection start date can be passed as a parameter 1`] = `
Array [
  <h3
    className="ecl-u-type-heading-3"
  >
    Initiative progress
  </h3>,
  <ol
    className="ecl-timeline"
    data-ecl-timeline="true"
  >
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Registered
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Collection start date
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item ecl-timeline__item--current"
    >
      <div
        className="ecl-timeline__label"
      >
        Ongoing
      </div>
      <div
        className="ecl-timeline__content"
      >
        09/09/2019
      </div>
    </li>
  </ol>,
]
`;

exports[`Progress collection start date can be passed as a step 1`] = `
Array [
  <h3
    className="ecl-u-type-heading-3"
  >
    Initiative progress
  </h3>,
  <ol
    className="ecl-timeline"
    data-ecl-timeline="true"
  >
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Registered
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Collection start date
      </div>
      <div
        className="ecl-timeline__content"
      >
        22/09/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item ecl-timeline__item--current"
    >
      <div
        className="ecl-timeline__label"
      >
        Ongoing
      </div>
      <div
        className="ecl-timeline__content"
      >
        09/09/2019
      </div>
    </li>
  </ol>,
]
`;

exports[`Progress footnote is displayed given a flag for earlier closure date 1`] = `
Array [
  <h3
    className="ecl-u-type-heading-3"
  >
    Initiative progress
  </h3>,
  <ol
    className="ecl-timeline"
    data-ecl-timeline="true"
  >
    <li
      className="ecl-timeline__item ecl-timeline__item--current"
    >
      <div
        className="ecl-timeline__label"
      >
        Ongoing
      </div>
      <div
        className="ecl-timeline__content"
      >
        09/09/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Closed
        <span
          className="ecl-u-type-color-red"
        >
           *
        </span>
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
  </ol>,
  <p
    class="ecl-u-type-paragraph-s ecl-u-type-color-red"
  >
    Collection closed earlier by organisers.
  </p>,
]
`;

exports[`Progress no input: provides feedback for unsupported timeline 1`] = `
Array [
  <h3
    className="ecl-u-type-heading-3"
  >
    Initiative progress
  </h3>,
  <p
    className="ecl-u-type-paragraph"
  >
    Given status cannot be mapped to currently supported timeline.
  </p>,
]
`;

exports[`Progress renders timeline in right order regardless of input order 1`] = `
Array [
  <h3
    className="ecl-u-type-heading-3"
  >
    Initiative progress
  </h3>,
  <ol
    className="ecl-timeline"
    data-ecl-timeline="true"
  >
    <li
      className="ecl-timeline__item"
    >
      <div
        className="ecl-timeline__label"
      >
        Registered
      </div>
      <div
        className="ecl-timeline__content"
      >
        11/08/2019
      </div>
    </li>
    <li
      className="ecl-timeline__item ecl-timeline__item--current"
    >
      <div
        className="ecl-timeline__label"
      >
        Ongoing
      </div>
      <div
        className="ecl-timeline__content"
      >
        09/09/2019
      </div>
    </li>
  </ol>,
]
`;

exports[`Progress utilities steps supported are public 1`] = `
Array [
  "REGISTERED",
  "COLLECTION_START_DATE",
  "ONGOING",
  "COLLECTION_EARLY_CLOSURE",
  "CLOSED",
  "VERIFICATION",
  "SUBMITTED",
  "ANSWERED",
]
`;
